File: blind75/64_validate_binary_search_tree/validate_bst.py
from typing import Optional


class TreeNode:
    """Represents a node in a binary tree."""

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def is_valid_bst(root: Optional[TreeNode]) -> bool:
    """
    Determines if a binary tree is a valid BST.
    Uses recursive range validation for O(n) time complexity.

    Args:
        root: Root node of the binary tree

    Returns:
        True if the tree is a valid BST, False otherwise
    """

    def validate(node: Optional[TreeNode], min_val: float, max_val: float) -> bool:
        """
        Validates that the node's value is within the valid range
        and recursively validates left and right subtrees.
        """
        if node is None:
            return True

        # Check if current node's value is within valid range
        if node.val <= min_val or node.val >= max_val:
            return False

        # Validate left subtree (all values must be < node.val)
        # and right subtree (all values must be > node.val)
        return validate(node.left, min_val, node.val) and validate(
            node.right, node.val, max_val
        )

    return validate(root, float("-inf"), float("inf"))

File: blind75/64_validate_binary_search_tree/validate_bst.go
package validate_bst

import "math"

// TreeNode represents a node in a binary tree.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// IsValidBST determines if a binary tree is a valid BST.
// Uses recursive range validation for O(n) time complexity.
func IsValidBST(root *TreeNode) bool {
	return isValidBSTHelper(root, math.MinInt64, math.MaxInt64)
}

// isValidBSTHelper validates that the node's value is within the valid range
// and recursively validates left and right subtrees with updated ranges.
func isValidBSTHelper(node *TreeNode, min, max int64) bool {
	if node == nil {
		return true
	}

	// Check if current node's value is within valid range
	if int64(node.Val) <= min || int64(node.Val) >= max {
		return false
	}

	// Validate left subtree (all values must be < node.Val)
	// and right subtree (all values must be > node.Val)
	return isValidBSTHelper(node.Left, min, int64(node.Val)) &&
		isValidBSTHelper(node.Right, int64(node.Val), max)
}
